src/control/tools/SplineHandler.cpp

```diff
@@ -0,0 +1,210 @@
+#include "SplineHandler.h"
+
+#include <cmath>
+
+#include "control/Control.h"
+#include "control/layer/LayerController.h"
+#include "gui/XournalView.h"
+#include "gui/XournalppCursor.h"
+#include "undo/InsertUndoAction.h"
+#include "util/cpp14memory.h"
+
+/**
+ * @brief A class to handle splines
+ *
+ * Drawing of a spline is started by a ButtonPressEvent. After every ButtonReleaseEvent,
+ * a new knot point is added. The spline is finished through a ButtonDoublePressEvent,
+ * the escape key or a ButtonPressEvent near the first knot. The latter event closes the spline.
+ *
+ * Splines segments can be linear or cubic (as in Inkscape). Join of two cubic segments
+ * is supposed to be smooth. As for now only linear splines are implemented.
+ */
+SplineHandler::SplineHandler(XournalView* xournal, XojPageView* redrawable, const PageRef& page):
+        InputHandler(xournal, redrawable, page) {}
+
+SplineHandler::~SplineHandler() = default;
+
+void SplineHandler::draw(cairo_t* cr) {
+    if (stroke) {
+        double zoom = xournal->getZoom();
+        int dpiScaleFactor = xournal->getDpiScaleFactor();
+
+        cairo_scale(cr, zoom * dpiScaleFactor, zoom * dpiScaleFactor);
+        // draw circles around knot points
+        double lineWidth = 0.5;
```

Comment:
In my opinion, the circles should scale with the zoom; they're difficult to see on larger screens.

Reply:
This is now done.

Reply:
Oops, I wasn't clear enough. It should scale _inversely_ with zoom, such that if the zoom is decreased, the circles will increase in size (so that they will still be visible).

Reply:
I have changed that accordingly now. Please check if this looks good on higher resolution screens now.